server/raft.go

```diff
@@ -1160,13 +1161,39 @@ func (n *raft) isCatchingUp() bool {
 	return n.catchup != nil
 }
 
-// Lock should be held.
+// Lock should be held. This function may block for up to ~5ms to check
+// forward progress in some cases.
 func (n *raft) isCurrent() bool {
-	// First check if have had activity and we match commit and applied.
-	if n.commit == 0 || n.commit != n.applied {
-		n.debug("Not current, commit %d != applied %d", n.commit, n.applied)
-		return false
+	// First check if have had activity and that commit and applied are equal.
+	// If not, wait for a short period of time and see if we are making forward
+	// progress.
+	startCommit, startDelta := n.commit, n.commit-n.applied
+	if startCommit == 0 || startDelta > 0 {
+		var progress bool
+		for i := 0; i < 10; i++ { // 5ms, in 0.5ms increments
+			n.RUnlock()
+			time.Sleep(time.Millisecond / 2)
+			n.RLock()
+			curCommit, curDelta := n.commit, n.commit-n.applied
```

Comment:
I think this can be simplified to a simple, is the delta smaller, wdyt?

Just track delta and make sure it descreases.

Reply:
Originally I checked that the delta was getting smaller only, but then if you have the same number of items coming in as being processed, it doesn't look as though anything is being done. By being able to check whether the commit has increased as well, we can at least check if forward progress is being made, even if the gap isn't getting smaller. I'm not sure if that's something we want to consider "healthy" though?

Reply:
If the delta remains same or increases that is not healthy imo. We are not catching up and should return false. So I think healthy is we are caught up, applied == committed, or we know delta is decreasing.